Error in slotted runtime for complex OPTIONAL MATCH

As a followup to #10910, I changed to Enterprise 3.4.0.alpha08 and continued testing the new runtime.

- Neo4j version: 3.4.0.alpha08 enterprise
- Operating system: Ubuntu 16.04
- API/Driver: any

I was experimenting with LDBC queries and ran into an error with [BI query 25](https://github.com/ldbc/ldbc_snb_implementations/blob/e3e08a1a72bb74d81d183e8125eee55f8585aa7c/cypher/queries-opencypher/bi-25.cypher). I managed to trim down the query to this:

```
MATCH path=allShortestPaths((p1:Person {id: 17592186045594})-[:KNOWS*]-(p2:Person {id: 2199023256530}))
UNWIND relationships(path) AS k
WITH path, startNode(k) AS pA, endNode(k) AS pB

OPTIONAL MATCH
  (pA)<-[:HAS_CREATOR]-(c:Comment)-[:REPLY_OF]->(post:Post)-[:HAS_CREATOR]->(pB),
  (post)<-[:CONTAINER_OF]-(forum:Forum)
WHERE forum.creationDate >= 20110201000000000 AND forum.creationDate <= 20110301000000000

RETURN *
```

The error message on the UI is:

> Neo.DatabaseError.General.UnknownError
> assertion failed

The `debug.log` shows the full stacktrace:

```
java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:156)
	at org.neo4j.cypher.internal.runtime.slotted.SlottedExecutionContext$$anonfun$mergeWith$1.apply(SlottedExecutionContext.scala:216)
	at org.neo4j.cypher.internal.runtime.slotted.SlottedExecutionContext$$anonfun$mergeWith$1.apply(SlottedExecutionContext.scala:210)
	at scala.collection.mutable.HashMap$$anonfun$foreach$1.apply(HashMap.scala:130)
	at scala.collection.mutable.HashMap$$anonfun$foreach$1.apply(HashMap.scala:130)
	at scala.collection.mutable.HashTable$class.foreachEntry(HashTable.scala:236)
	at scala.collection.mutable.HashMap.foreachEntry(HashMap.scala:40)
	at scala.collection.mutable.HashMap.foreach(HashMap.scala:130)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.SlotConfiguration.foreachSlot(SlotConfiguration.scala:298)
	at org.neo4j.cypher.internal.runtime.slotted.SlottedExecutionContext.mergeWith(SlottedExecutionContext.scala:210)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.NodeRightOuterHashJoinPipe$$anonfun$internalCreateResults$1$$anonfun$apply$1.apply(NodeRightOuterHashJoinPipe.scala:43)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.NodeRightOuterHashJoinPipe$$anonfun$internalCreateResults$1$$anonfun$apply$1.apply(NodeRightOuterHashJoinPipe.scala:43)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.LinearSeqOptimized$class.foreach(LinearSeqOptimized.scala:73)
	at scala.collection.mutable.MutableList.foreach(MutableList.scala:30)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.NodeRightOuterHashJoinPipe$$anonfun$internalCreateResults$1.apply(NodeRightOuterHashJoinPipe.scala:43)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.NodeRightOuterHashJoinPipe$$anonfun$internalCreateResults$1.apply(NodeRightOuterHashJoinPipe.scala:37)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.TraversableOnce$FlattenOps$$anon$1.hasNext(TraversableOnce.scala:464)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.ClosingIterator$$anonfun$hasNext$1.apply$mcZ$sp(ResultIterator.scala:57)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.ClosingIterator$$anonfun$hasNext$1.apply(ResultIterator.scala:55)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.ClosingIterator$$anonfun$hasNext$1.apply(ResultIterator.scala:55)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.ClosingIterator$$anonfun$failIfThrows$1.apply(ResultIterator.scala:82)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.ClosingIterator.decoratedCypherException(ResultIterator.scala:91)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.ClosingIterator.failIfThrows(ResultIterator.scala:80)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.ClosingIterator.hasNext(ResultIterator.scala:54)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
	at org.neo4j.cypher.internal.runtime.RuntimeJavaValueConverter$feedIteratorToVisitable.accept(RuntimeJavaValueConverter.scala:53)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.PipeExecutionResult.accept(PipeExecutionResult.scala:112)
	at org.neo4j.cypher.internal.compatibility.ClosingExecutionResult$$anonfun$accept$2.apply$mcV$sp(ClosingExecutionResult.scala:174)
	at org.neo4j.cypher.internal.compatibility.ClosingExecutionResult$$anonfun$accept$2.apply(ClosingExecutionResult.scala:173)
	at org.neo4j.cypher.internal.compatibility.ClosingExecutionResult$$anonfun$accept$2.apply(ClosingExecutionResult.scala:173)
	at org.neo4j.cypher.exceptionHandler$runSafely$.apply(exceptionHandler.scala:88)
	at org.neo4j.cypher.internal.compatibility.ClosingExecutionResult.accept(ClosingExecutionResult.scala:173)
	at org.neo4j.bolt.v1.runtime.CypherAdapterStream.accept(CypherAdapterStream.java:76)
	at org.neo4j.bolt.v1.messaging.BoltMessageRouter$ResultHandler.onRecords(BoltMessageRouter.java:146)
	at org.neo4j.bolt.v1.runtime.BoltStateMachine$State$3.lambda$pullAll$0(BoltStateMachine.java:454)
	at org.neo4j.bolt.v1.runtime.TransactionStateMachine$State$1.streamResult(TransactionStateMachine.java:267)
	at org.neo4j.bolt.v1.runtime.TransactionStateMachine.streamResult(TransactionStateMachine.java:104)
	at org.neo4j.bolt.v1.runtime.BoltStateMachine$State$3.pullAll(BoltStateMachine.java:453)
	at org.neo4j.bolt.v1.runtime.BoltStateMachine.pullAll(BoltStateMachine.java:236)
	at org.neo4j.bolt.v1.messaging.BoltMessageRouter.lambda$onPullAll$6(BoltMessageRouter.java:114)
	at org.neo4j.bolt.v1.runtime.MonitoredWorkerFactory$MonitoredBoltWorker.lambda$enqueue$0(MonitoredWorkerFactory.java:82)
	at org.neo4j.bolt.v1.runtime.concurrent.RunnableBoltWorker.execute(RunnableBoltWorker.java:150)
	at org.neo4j.bolt.v1.runtime.concurrent.RunnableBoltWorker.executeBatch(RunnableBoltWorker.java:143)
	at org.neo4j.bolt.v1.runtime.concurrent.RunnableBoltWorker.run(RunnableBoltWorker.java:108)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at org.neo4j.helpers.NamedThreadFactory$2.run(NamedThreadFactory.java:110)
```

The assertion error comes from https://github.com/neo4j/neo4j/blob/9348d2e7a3dfe1379e461ea9164f49d961df6cc3/enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/SlottedExecutionContext.scala#L216

Similarly to the previous issue, prefixing the query with `CYPHER runtime=interpreted` makes the error go away, so this must be an issue with the compiled runtime.

Reproducing the error needs some data. The scale factor 0.1 data set of the LDBC is sufficient for this goal:
* [zipped `graph.db` (Dropbox)](https://www.dropbox.com/s/z8l7o0yellsattr/graph.db.zip?dl=0)
* [CSV files with import instructions (GitHub)](https://github.com/szarnyasg/ldbc-sf01)

I am cc-ing @alexaverbuch who helped me debug the cause of this issue.

--- a/community/cypher/util-3.4/src/main/java/org/neo4j/cypher/internal/util/v3_4/AssertionRunner.java
+++ b/community/cypher/util-3.4/src/main/java/org/neo4j/cypher/internal/util/v3_4/AssertionRunner.java
@@ -46,4 +46,11 @@ public class AssertionRunner
     {
         void apply();
     }
+
+    public static boolean isAssertionsEnabled()
+    {
+        boolean assertionsEnabled = false;
+        assert assertionsEnabled = true;
+        return assertionsEnabled;
+    }
 }

--- a/community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/AssertionUtils.scala
+++ b/community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/AssertionUtils.scala
@@ -24,4 +24,6 @@ object AssertionUtils {
       override def apply() = f
     })
   }
+
+  def assertionsEnabled = AssertionRunner.isAssertionsEnabled
 }

--- a/enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/JoinAcceptanceTest.scala
+++ b/enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/JoinAcceptanceTest.scala
@@ -89,6 +89,44 @@ class JoinAcceptanceTest extends ExecutionEngineFunSuite with CypherComparisonSu
       planComparisonStrategy = ComparePlansWithAssertion(_ should useOperators("NodeRightOuterHashJoin"), Configs.AllRulePlanners + Configs.BackwardsCompatibility))
   }
 
+  test("should handle node left outer hash join with different types for the node variable") {
+    val a = createLabeledNode(Map[String, Any]("name" -> "a"), "A")
+    createLabeledNode(Map[String, Any]("name" -> "a2"), "A")
+    for(i <- 0 until 50) { // This number is sensitive in that it has to exceed the cardinality estimation of the UNWIND
+      val b = createLabeledNode(Map[String, Any]("name" -> s"${i}b"), "B")
+      if(i != 0) relate(a, b)
+    }
+
+    val query = """MATCH (a:A)
+                  |UNWIND [a] as refA
+                  |OPTIONAL MATCH (refA)-->(b:B)
+                  |USING JOIN ON refA
+                  |RETURN a.name, b.name""".stripMargin
+
+    val expectSucceed = Configs.Interpreted - Configs.Cost2_3 - Configs.Cost3_1
+    executeWith(expectSucceed, query,
+      planComparisonStrategy = ComparePlansWithAssertion(_ should useOperators("NodeLeftOuterHashJoin"), expectPlansToFail))
+  }
+
+  test("should handle node right outer hash join with different types for the node variable") {
+    val b = createLabeledNode(Map[String, Any]("name" -> "b"), "B")
+    createLabeledNode(Map[String, Any]("name" -> "b2"), "B")
+    for(i <- 0 until 10) {
+      val a = createLabeledNode(Map[String, Any]("name" -> s"${i}a"), "A")
+      if(i == 0) relate(a, b)
+    }
+
+    val query = """MATCH (a:A)
+                  |UNWIND [a] as refA
+                  |OPTIONAL MATCH (refA)-->(b:B)
+                  |USING JOIN ON refA
+                  |RETURN a.name, b.name""".stripMargin
+
+    val expectSucceed = Configs.Interpreted - Configs.Cost2_3 - Configs.Cost3_1
+    val result = executeWith(expectSucceed, query,
+      planComparisonStrategy = ComparePlansWithAssertion(_ should useOperators("NodeRightOuterHashJoin"), Configs.AllRulePlanners + Configs.BackwardsCompatibility))
+  }
+
   test("optional match join should not crash") {
     val query =
       """MATCH (a:A)-->(b:B)-->(c:C)

--- a/enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/SlotConfiguration.scala
+++ b/enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/SlotConfiguration.scala
@@ -160,6 +160,8 @@ class SlotConfiguration(private val slots: mutable.Map[String, Slot],
 
   private val getters: mutable.Map[String, ExecutionContext => AnyValue] = new mutable.HashMap[String, ExecutionContext => AnyValue]()
   private val setters: mutable.Map[String, (ExecutionContext, AnyValue) => Unit] = new mutable.HashMap[String, (ExecutionContext, AnyValue) => Unit]()
+  private val primitiveNodeSetters: mutable.Map[String, (ExecutionContext, Long) => Unit] = new mutable.HashMap[String, (ExecutionContext, Long) => Unit]()
+  private val primitiveRelationshipSetters: mutable.Map[String, (ExecutionContext, Long) => Unit] = new mutable.HashMap[String, (ExecutionContext, Long) => Unit]()
 
   def size() = SlotConfiguration.Size(numberOfLongs, numberOfReferences)
 
@@ -272,9 +274,13 @@ class SlotConfiguration(private val slots: mutable.Map[String, Slot],
     case _ => throw new InternalException(s"Uh oh... There was no slot for `$name`")
   }
 
-  def updateAccessorFunctions(key: String, getter: ExecutionContext => AnyValue, setter: (ExecutionContext, AnyValue) => Unit) = {
+  def updateAccessorFunctions(key: String, getter: ExecutionContext => AnyValue, setter: (ExecutionContext, AnyValue) => Unit,
+                              primitiveNodeSetter: Option[(ExecutionContext, Long) => Unit],
+                              primitiveRelationshipSetter: Option[(ExecutionContext, Long) => Unit]) = {
     getters += key -> getter
     setters += key -> setter
+    primitiveNodeSetter.map(primitiveNodeSetters += key -> _)
+    primitiveRelationshipSetter.map(primitiveRelationshipSetters += key -> _)
   }
 
   def getter(key: String): ExecutionContext => AnyValue = {
@@ -291,6 +297,14 @@ class SlotConfiguration(private val slots: mutable.Map[String, Slot],
 
   def maybeSetter(key: String): Option[(ExecutionContext, AnyValue) => Unit] = {
     setters.get(key)
+  }
+
+  def maybePrimitiveNodeSetter(key: String): Option[(ExecutionContext, Long) => Unit] = {
+    primitiveNodeSetters.get(key)
+  }
+
+  def maybePrimitiveRelationshipSetter(key: String): Option[(ExecutionContext, Long) => Unit] = {
+    primitiveRelationshipSetters.get(key)
   }
 
   // NOTE: This will give duplicate slots when we have aliases

--- a/enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/SlottedExecutionContext.scala
+++ b/enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/SlottedExecutionContext.scala
@@ -22,6 +22,7 @@ package org.neo4j.cypher.internal.runtime.slotted
 import org.neo4j.cypher.internal.compatibility.v3_4.runtime.{LongSlot, RefSlot, SlotConfiguration}
 import org.neo4j.cypher.internal.runtime.interpreted.ExecutionContext
 import org.neo4j.cypher.internal.runtime.slotted.helpers.NullChecker.entityIsNull
+import org.neo4j.cypher.internal.util.v3_4.AssertionUtils._
 import org.neo4j.cypher.internal.util.v3_4.InternalException
 import org.neo4j.cypher.internal.util.v3_4.symbols.{CTNode, CTRelationship}
 import org.neo4j.values.AnyValue
@@ -206,21 +207,37 @@ case class SlottedExecutionContext(slots: SlotConfiguration) extends ExecutionCo
     refs(offset)
 
   override def mergeWith(other: ExecutionContext): ExecutionContext = other match {
-    case primitiveOther: SlottedExecutionContext =>
-      primitiveOther.slots.foreachSlot {
-        case (key, otherSlot) =>
-          val thisSlot = slots.get(key).getOrElse(
+    case slottedOther: SlottedExecutionContext =>
+      slottedOther.slots.foreachSlot {
+        case (key, otherSlot @ LongSlot(offset, _, CTNode)) =>
+          val thisSlotSetter = slots.maybePrimitiveNodeSetter(key).getOrElse(
+            throw new InternalException(s"Tried to merge primitive node slot $otherSlot from $other but it is missing from $this." +
+              "Looks like something needs to be fixed in slot allocation.")
+          )
+          thisSlotSetter.apply(this, other.getLongAt(offset))
+
+        case (key, otherSlot @ LongSlot(offset, _, CTRelationship)) =>
+          val thisSlotSetter = slots.maybePrimitiveRelationshipSetter(key).getOrElse(
+            throw new InternalException(s"Tried to merge primitive relationship slot $otherSlot from $other but it is missing from $this." +
+              "Looks like something needs to be fixed in slot allocation.")
+          )
+          thisSlotSetter.apply(this, other.getLongAt(offset))
+
+        case (key, otherSlot @ RefSlot(offset, _, _)) if slottedOther.isRefInitialized(offset) =>
+          val thisSlotSetter = slots.maybeSetter(key).getOrElse(
             throw new InternalException(s"Tried to merge slot $otherSlot from $other but it is missing from $this." +
               "Looks like something needs to be fixed in slot allocation.")
           )
-          assert(thisSlot.isTypeCompatibleWith(otherSlot))
-          otherSlot match {
-            case LongSlot(offset, _, _) =>
-              setLongAt(thisSlot.offset, other.getLongAt(offset))
-            case RefSlot(offset, _, _) if primitiveOther.isRefInitialized(offset) =>
-              setRefAt(thisSlot.offset, primitiveOther.getRefAtWithoutCheckingInitialized(offset))
-            case _ =>
+
+          ifAssertionsEnabled {
+            val thisSlot = slots.get(key).get
+            // This should be guaranteed by slot allocation or else we could get incorrect results
+            if (!thisSlot.nullable && otherSlot.nullable)
+              throw new InternalException(s"Tried to merge slot $otherSlot into $thisSlot but its nullability is incompatible")
           }
+
+          val otherValue = slottedOther.getRefAtWithoutCheckingInitialized(offset)
+          thisSlotSetter.apply(this, otherValue)
       }
       this
 

--- a/enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/SlottedPipeBuilder.scala
+++ b/enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/SlottedPipeBuilder.scala
@@ -99,7 +99,18 @@ class SlottedPipeBuilder(fallback: PipeBuilder,
       case (key, slot) =>
         val getter = SlottedPipeBuilderUtils.makeGetValueFromSlotFunctionFor(slot)
         val setter = SlottedPipeBuilderUtils.makeSetValueInSlotFunctionFor(slot)
-        slots.updateAccessorFunctions(key, getter, setter)
+        val primitiveNodeSetter =
+          if (slot.typ.isAssignableFrom(CTNode))
+            Some(SlottedPipeBuilderUtils.makeSetPrimitiveNodeInSlotFunctionFor(slot))
+          else
+            None
+        val primitiveRelationshipSetter =
+          if (slot.typ.isAssignableFrom(CTRelationship))
+            Some(SlottedPipeBuilderUtils.makeSetPrimitiveRelationshipInSlotFunctionFor(slot))
+          else
+            None
+
+       slots.updateAccessorFunctions(key, getter, setter, primitiveNodeSetter, primitiveRelationshipSetter)
     }
   }
 

--- a/enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/helpers/SlottedPipeBuilderUtils.scala
+++ b/enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/helpers/SlottedPipeBuilderUtils.scala
@@ -22,11 +22,11 @@ package org.neo4j.cypher.internal.runtime.slotted.helpers
 import NullChecker.entityIsNull
 import org.neo4j.cypher.internal.compatibility.v3_4.runtime.{LongSlot, RefSlot, Slot}
 import org.neo4j.cypher.internal.runtime.interpreted.ExecutionContext
-import org.neo4j.cypher.internal.util.v3_4.{InternalException, ParameterWrongTypeException}
+import org.neo4j.cypher.internal.util.v3_4.{AssertionUtils, InternalException, ParameterWrongTypeException}
 import org.neo4j.cypher.internal.util.v3_4.symbols.{CTNode, CTRelationship, CypherType}
 import org.neo4j.values.AnyValue
 import org.neo4j.values.storable.Values
-import org.neo4j.values.virtual.{VirtualRelationshipValue, VirtualNodeValue, VirtualValues}
+import org.neo4j.values.virtual.{VirtualNodeValue, VirtualRelationshipValue, VirtualValues}
 
 object SlottedPipeBuilderUtils {
   // TODO: Check if having try/catch blocks inside some of these generated functions prevents inlining or other JIT optimizations
@@ -201,4 +201,94 @@ object SlottedPipeBuilderUtils {
       case _ =>
         throw new InternalException(s"Do not know how to make setter for slot $slot")
     }
+
+  /**
+    * Use this to make a specialized setter function for a slot,
+    * that takes as input an ExecutionContext and a primitive long value.
+    */
+  def makeSetPrimitiveInSlotFunctionFor(slot: Slot, valueType: CypherType): (ExecutionContext, Long) => Unit =
+    (slot, valueType) match {
+      case (LongSlot(offset, nullable, CTNode), CTNode) =>
+        if (AssertionUtils.assertionsEnabled && !nullable) {
+          (context: ExecutionContext, value: Long) =>
+            if (value == -1L)
+              throw new ParameterWrongTypeException(s"Cannot assign null to a non-nullable slot")
+            context.setLongAt(offset, value)
+        }
+        else {
+          (context: ExecutionContext, value: Long) =>
+            context.setLongAt(offset, value)
+        }
+
+      case (LongSlot(offset, nullable, CTRelationship), CTRelationship) =>
+        if (AssertionUtils.assertionsEnabled && !nullable) {
+          (context: ExecutionContext, value: Long) =>
+            if (value == -1L)
+              throw new ParameterWrongTypeException(s"Cannot assign null to a non-nullable slot")
+            context.setLongAt(offset, value)
+        }
+        else {
+          (context: ExecutionContext, value: Long) =>
+            context.setLongAt(offset, value)
+        }
+
+      case (RefSlot(offset, false, typ), CTNode) if typ.isAssignableFrom(CTNode) =>
+        if (AssertionUtils.assertionsEnabled) {
+          (context: ExecutionContext, value: Long) =>
+            if (value == -1L)
+              throw new ParameterWrongTypeException(s"Cannot assign null to a non-nullable slot")
+            context.setRefAt(offset, VirtualValues.node(value))
+        }
+        else {
+          (context: ExecutionContext, value: Long) =>
+            // NOTE: Slot allocation needs to guarantee that we can never get nulls in here
+            context.setRefAt(offset, VirtualValues.node(value))
+        }
+
+      case (RefSlot(offset, false, typ), CTRelationship) if typ.isAssignableFrom(CTRelationship) =>
+        if (AssertionUtils.assertionsEnabled) {
+          (context: ExecutionContext, value: Long) =>
+            if (value == -1L)
+              throw new ParameterWrongTypeException(s"Cannot assign null to a non-nullable slot")
+            context.setRefAt(offset, VirtualValues.relationship(value))
+        }
+        else {
+          (context: ExecutionContext, value: Long) =>
+            // NOTE: Slot allocation needs to guarantee that we can never get nulls in here
+            context.setRefAt(offset, VirtualValues.relationship(value))
+        }
+
+      case (RefSlot(offset, true, typ), CTNode) if typ.isAssignableFrom(CTNode) =>
+        (context: ExecutionContext, value: Long) =>
+          if (value == -1L)
+            context.setRefAt(offset, Values.NO_VALUE)
+          else
+            context.setRefAt(offset, VirtualValues.node(value))
+
+      case (RefSlot(offset, true, typ), CTRelationship) if typ.isAssignableFrom(CTRelationship) =>
+        (context: ExecutionContext, value: Long) =>
+          if (value == -1L)
+            context.setRefAt(offset, Values.NO_VALUE)
+          else
+            context.setRefAt(offset, VirtualValues.relationship(value))
+
+      case _ =>
+        throw new InternalException(s"Do not know how to make a primitive $valueType setter for slot $slot")
+    }
+
+  /**
+    * Use this to make a specialized getter function for a slot that is expected to contain a node
+    * that given an ExecutionContext returns a long with the node id.
+    */
+  def makeSetPrimitiveNodeInSlotFunctionFor(slot: Slot): (ExecutionContext, Long) => Unit =
+    makeSetPrimitiveInSlotFunctionFor(slot, CTNode)
+
+  /**
+    * Use this to make a specialized getter function for a slot that is expected to contain a node
+    * that given an ExecutionContext returns a long with the relationship id.
+    */
+  def makeSetPrimitiveRelationshipInSlotFunctionFor(slot: Slot): (ExecutionContext, Long) => Unit =
+    makeSetPrimitiveInSlotFunctionFor(slot, CTRelationship)
+
+
 }

--- a/enterprise/cypher/slotted-runtime/src/test/scala/org/neo4j/cypher/internal/runtime/slotted/helpers/SlottedPipeBuilderUtilsTest.scala
+++ b/enterprise/cypher/slotted-runtime/src/test/scala/org/neo4j/cypher/internal/runtime/slotted/helpers/SlottedPipeBuilderUtilsTest.scala
@@ -22,6 +22,7 @@ package org.neo4j.cypher.internal.runtime.slotted.helpers
 import SlottedPipeBuilderUtils._
 import org.neo4j.cypher.internal.compatibility.v3_4.runtime.{Slot, SlotConfiguration}
 import org.neo4j.cypher.internal.runtime.slotted.SlottedExecutionContext
+import org.neo4j.cypher.internal.util.v3_4.AssertionUtils._
 import org.neo4j.cypher.internal.util.v3_4.ParameterWrongTypeException
 import org.neo4j.cypher.internal.util.v3_4.symbols._
 import org.neo4j.cypher.internal.util.v3_4.test_helpers.CypherFunSuite
@@ -37,6 +38,7 @@ class SlottedPipeBuilderUtilsTest extends CypherFunSuite {
     .newLong("r1", false, CTRelationship)
     .newLong("r2", true, CTRelationship)
     .newReference("x", true, CTAny)
+    .newReference("y", false, CTAny)
 
   // GETTING
 
@@ -158,5 +160,113 @@ class SlottedPipeBuilderUtilsTest extends CypherFunSuite {
     setter(context, expectedValue)
     val value = context.getRefAt(slot.offset)
     value should equal(expectedValue)
+  }
+
+  private def assertPrimitiveNodeSetLong(slot: Slot, id: Long): Unit = {
+    val context = SlottedExecutionContext(slots)
+    val primitiveNodeSetter = makeSetPrimitiveNodeInSlotFunctionFor(slot)
+
+    primitiveNodeSetter(context, id)
+    context.getLongAt(slot.offset) should equal(id)
+  }
+
+  private def assertPrimitiveRelationshipSetLong(slot: Slot, id: Long): Unit = {
+    val context = SlottedExecutionContext(slots)
+    val primitiveRelationshipSetter = makeSetPrimitiveRelationshipInSlotFunctionFor(slot)
+
+    primitiveRelationshipSetter(context, id)
+    context.getLongAt(slot.offset) should equal(id)
+  }
+
+  private def assertPrimitiveNodeSetRef(slot: Slot, id: Long, expected: AnyValue): Unit = {
+    val context = SlottedExecutionContext(slots)
+    val primitiveNodeSetter = makeSetPrimitiveNodeInSlotFunctionFor(slot)
+
+    primitiveNodeSetter(context, id)
+    context.getRefAt(slot.offset) should equal(expected)
+  }
+
+  private def assertPrimitiveRelationshipSetRef(slot: Slot, id: Long, expected: AnyValue): Unit = {
+    val context = SlottedExecutionContext(slots)
+    val primitiveRelationshipSetter = makeSetPrimitiveRelationshipInSlotFunctionFor(slot)
+
+    primitiveRelationshipSetter(context, id)
+    context.getRefAt(slot.offset) should equal(expected)
+  }
+
+  private def assertPrimitiveNodeSetFails(slot: Slot, id: Long): Unit = {
+    val context = SlottedExecutionContext(slots)
+    val setter = makeSetPrimitiveNodeInSlotFunctionFor(slot)
+
+    // The setter only throws if assertions are enabled
+    ifAssertionsEnabled {
+      a[ParameterWrongTypeException] should be thrownBy (setter(context, id))
+    }
+  }
+
+  private def assertPrimitiveRelationshipSetFails(slot: Slot, id: Long): Unit = {
+    val context = SlottedExecutionContext(slots)
+    val setter = makeSetPrimitiveRelationshipInSlotFunctionFor(slot)
+
+    // The setter only throws if assertions are enabled
+    ifAssertionsEnabled {
+      a[ParameterWrongTypeException] should be thrownBy (setter(context, id))
+    }
+  }
+
+  test("primitive node setter for non-nullable node slot") {
+    assertPrimitiveNodeSetLong(slots("n1"), 42L)
+  }
+
+  test("primitive node setter for nullable node slots with null") {
+    assertPrimitiveNodeSetLong(slots("n2"), -1L)
+  }
+
+  test("primitive node setter for nullable node slot") {
+    assertPrimitiveNodeSetLong(slots("n2"), 42L)
+  }
+
+  test("primitive relationship setter for non-nullable relationship slot") {
+    assertPrimitiveRelationshipSetLong(slots("r1"), 42L)
+  }
+
+  test("primitive relationship setter for nullable relationship slots with null") {
+    assertPrimitiveRelationshipSetLong(slots("r2"), -1)
+  }
+
+  test("primitive relationship setter for nullable relationship slot") {
+    assertPrimitiveRelationshipSetLong(slots("r2"), 42L)
+  }
+
+  test("primitive node setter for non-nullable node slot should throw") {
+    assertPrimitiveNodeSetFails(slots("n1"), -1L)
+  }
+
+  test("primitive relationship setter for non-nullable relationship slot should throw") {
+    assertPrimitiveRelationshipSetFails(slots("r1"), -1L)
+  }
+
+  test("primitive node setter for ref slot") {
+    assertPrimitiveNodeSetRef(slots("x"), 42L, VirtualValues.node(42L))
+  }
+
+  test("primitive node setter for nullable ref slot with null") {
+    assertPrimitiveNodeSetRef(slots("x"), -1L, Values.NO_VALUE)
+  }
+
+  test("primitive node setter for non-nullable ref slot should throw") {
+    assertPrimitiveNodeSetFails(slots("y"), -1L)
+  }
+
+  test("primitive relationship setter for ref slot") {
+    assertPrimitiveRelationshipSetRef(slots("x"), 42L, VirtualValues.relationship(42L))
+  }
+
+  test("primitive relationship setter for nullable ref slot with null") {
+    assertPrimitiveRelationshipSetRef(slots("x"), -1L, Values.NO_VALUE)
+  }
+
+  test("primitive relationship setter for non-nullable ref slot should throw") {
+    assertPrimitiveRelationshipSetFails(slots("y"), -1L)
   }
 }

